kafkabp/config.go

```diff
@@ -69,6 +72,15 @@ type ConsumerConfig struct {
 	// When GroupID is non-empty, Version must be at least "0.10.2.0".
 	GroupID string `yaml:"groupID"`
 
+	// Optional. This is only applicable when GroupID is empty string.
+	// When GroupID is empty, the configuration enables a TopicConsumer to read from
+	// all partitions of the given kafka stream. When PartitionsSelectorProvider is also specified
+	// the TopicConsumer will only consume partitions that evaluates to true with the given
+	// predicate returned from PartitionsSelectorProvider. If PartitionsSelectorProvider is specified,
+	// it must return a non-nil predicate of type PartitionsSelectorFunc, else the topic
+	// consumer will return an error of type ErrNilPartitionSelector
+	PartitionsSelectorProvider PartitionsSelectorProvider `yaml:"-"`
```

Comment:
Why the two levels of indirection? Why not just `func(paritionID, numPartitions int) bool`

Reply:
there could be expensive computation to generate `[]bool` (or a set of ints) from `numPartitions`, with `func(paritionID, numPartitions int) bool` you would need to use something like `sync.Once` to avoid doing that computation on each call. so tl;dr is that with two-layer approach we can shift all expensive work to this layer and make the next layer a simple lookup from either a slice or a set.

Reply:
Makse sense.  If we are expecting to cover the use-case where the provider is expensive, we should be clear about how ofren / when the provider is called.  My assumption was that this would basically be called each time as `provider(shards)(num)`, and without evidence I think a user should assume that too.